### PR TITLE
Update `upload-artifact` and `download-artifact`: one artifact per arch + platform tag series

### PR DESCRIPTION
**Pull request by gusmonod**
Another approach for #161 where we use one artifact per architecture + platform tag series instead of one per wheel.

### CODE REVIEW

**Comment by gusmonod** [comment]
@godlygeek this is now ready for review. Let me know what you think of this alternative approach.

File: .github/workflows/build_wheels.yml
Comment (godlygeek) [inline .github/workflows/build_wheels.yml:38]: Restore this before landing this change

File: .github/workflows/build_wheels.yml
Comment (godlygeek) [inline .github/workflows/build_wheels.yml:297]: And this

File: .github/workflows/build_wheels.yml
Comment (godlygeek) [inline .github/workflows/build_wheels.yml:306]: And this

**Review by godlygeek — CHANGES_REQUESTED** [review]
Marking as "request changes" just so we don't accidentally land this before putting some of the "only on releases" checks back.

**Comment by godlygeek** [comment]
Wow, OK.

First off, I think this looks a lot cleaner - both cleaner than #161, and also probably cleaner than what's on main.

It cuts the total amount of CPU time required for a full run (including aarch64 builds) down from 3.5 hours to 1.65 hours, which is borderline miraculous. It looks like the old approach was wasting a ton of time rebuilding `elfutils` from source over and over in a `before-all` step. This new approach only builds it once per architecture, which, yes, is what we ought to have done all along, and which saves an absolutely unbelievable amount of time when it comes to aarch64 builds. Those run under qemu, and are painfully slow, and the unnecessary repetitive builds of `elfutils` were using way more time than I realized.

It does mean that for a full run that includes the aarch64 builds, we can't test the x86-64 wheels until after the aarch64 ones have been built, which takes about an hour. That's a tad annoying, but oh well, that's why we don't build for aarch64 on every PR...

Building all the x86-64 wheels took 13 minutes, which means that on PRs there would be a 13 minute delay before the first tests start running. That's not _too_ terrible, but it looks like on main the wheels are are built in under 6 minutes thanks to the parallelism, so this would mean more than doubling the amount of time until we start to see test results. Maybe that's not so bad, though...

I think I'm comfortable landing this.

@pablogsal Any objections? The two downsides are
1. For releases, the tests won't run until after the aarch64 builds complete, which takes an hour.
2. For normal PRs, the tests will run as soon as all x86-64 wheels are built, but since we're sacrificing some parallelism, that will take a bit more than 2x as long as it used to.

In exchange for those tradeoffs, we reduce the total CPU time of a release run by more than half, we gain the ability to download the wheels while the tests are still running, and we get to upgrade from the v3 of `upload-artifact` and `download-artifact` to the v4.

I think that's a cost I'm willing to pay, but I want to make sure you're on board before landing this.

**Comment by godlygeek** [comment]
For the x86-64 wheels, perhaps we should do the manylinux and musllinux separately. Doing those in parallel should get probably us back to running the tests as quickly as we used to...

**Review by godlygeek — APPROVED** [review]
Alright @gusmonod - I've pushed two new commits, which re-add the checks for steps that happen only on releases, and which switch musllinux and manylinux builds to happen in parallel. With that, the time to run the test suite on a normal (non-release) PR run is back down to about what it is on main, including the latency before the first tests start running.

Check my work on the two commits I pushed. If you're happy with them, I'm happy landing this.

File: .github/workflows/build_wheels.yml
Comment (gusmonod) [inline .github/workflows/build_wheels.yml:37]: Why do you use an underscore (`_`) instead of a dash (`-`) for the separation between `manylinux` and `x86_64`? I would prefer a dash if possible, to make it clear that `x86_64` is really a single unit.

Comment (godlygeek) [reply .github/workflows/build_wheels.yml:37]: This gets expanded in `CIBW_BUILD: "cp3{7..12}-${{ matrix.wheel_type }}"` and cibuildwheel uses a `_` instead of a `-` there.  See https://cibuildwheel.readthedocs.io/en/stable/options/#build-skip for examples, but we need that to expand to (for instance) `cp312-musllinux_x86_64`

File: .github/workflows/build_wheels.yml
Comment (gusmonod) [inline .github/workflows/build_wheels.yml:48]: I would add `Build ` at the front there to make it clearer what this step does.

```suggestion
    name: Build ${{ matrix.wheel_type }} wheels
```

**Review by gusmonod — COMMENTED** [review]
This all looks good to me! I have one nitpick below, feel free to ignore it if you want to. As I have created the PR myself though, I cannot approve it.

File: .github/workflows/build_wheels.yml
Comment (godlygeek) [reply .github/workflows/build_wheels.yml]: That gets truncated in the UI:

image

Compared to without:

image

I think it looks better without the word "build", and I think the intention is clear enough given that it's in a box labeled "Matrix: build_wheels", so I think I'm gonna skip this suggestion.

Comment (gusmonod) [reply .github/workflows/build_wheels.yml:48]: I guess it depends on what you prioritize, between the UI in the PR checks (where adding `Build` probably makes more sense) or the GH actions summary view (where it's better left off). Either way, this was just a nitpick, it is indeed easy enough to understand.

**Pull request merged**
